packages/loader/container-loader/src/connectionManager.ts

```diff
@@ -530,6 +517,28 @@ export class ConnectionManager implements IConnectionManager {
 			requestedMode = "write";
 		}
 
+		if (this.connection !== undefined || this.pendingConnection !== undefined) {
+			// Connection attempt already completed successfully or is in progress
+			// In general, there should be no issues if the modes do not match:
+			// If at some point it was Ok to connect as "read" (i.e. there were no pending ops we had to track),
+			// then it should be Ok to use "read" connection even if for some reason request came in to connect as "write"
+			// (though that should never happen)
+			// The opposite should be fine as well: we may have had idle "write" connection, and request to reconnect came in,
+			// using default "read" mode.
+			// That all said, let's understand better where such mismatches are coming from.
+			const mode = this.connection?.mode ?? this.pendingConnection?.connectionMode;
+			if (mode !== requestedMode) {
+				this.logger.sendErrorEvent({
+					eventName: "ConnectionModeMismatch",
+					hasConnection: this.connection !== undefined,
```

Comment:
Renaming.
I do not see anywhere testing this.connection.disposed. I do see that we test connection.disposed when we receive connection, but that's mostly because we can get already disposed connection, before we had a chance to install event handlers. Otherwise, we learn about disposal through events and immediately get rid of connection. So if we have connection like that, it's not disposed (unless we have some other bug)